### PR TITLE
Feature/1696 honor memorial

### LINKED ISSUE #1696: Honor/Memorial Fields for Recurring Donations/Opportunities
This is a simple customization request, but working with Luminate again reminds me that they actually have thought this through a bit.  This same set of fields existing on both the Recurring Donation and Opportunity Objects (with pre-existing custom mapping, if this is even possible) would enable this quickly for folks without consulting partners needing to spend time building them.
honor memorial

### PR DESCRIPTION
**Pull request by njjc**
# Warning
# Info
- This release includes new Opportunity tribute fields to designate donations in honor and in memoriam.

# Issues
Fixes #1696

### CODE REVIEW

**Comment by kbromer** [comment]
**lurch: review apex visualforce

**Comment by kbromer** [comment]
**lurch: review

**Comment by kbromer** [comment]
**lurch: review apex visualforce

**Comment by LurchTheButler** [comment]
#### General
- [ ] All new pages, classes or metadata conform to naming conventions

#### Apex
- [ ] This request includes Apex
	- [ ] Sharing/Without/None has been explicitly determined and explained
	- [ ] FLS is appropriately checked and enforced
- [ ] Custom settings access is done through appropriate interfaces
- [ ] Methods are appropriately sized and scopes
- [ ] Inner classes utilized where appropriate
- [ ] Dependencies are done explicitly or have been noted where dynamic
- [ ] API Version is explicitly set to the latest or otherwise indicated as to why not
- [ ] Code is bulkified
	- [ ] No SOQL or DML in For loops
	- [ ] Async calls in For loops are properly rate limited
	- [ ] Code utilized DML wrapper where appropriate
- [ ] No new triggers have been added
- [ ] All currency operations are multi-currency capable and compliant
- [ ] Any new triggered functionality properly implements TDTM interface
	- [ ] A new TDTM record has been added to the install script for this class
	- [ ] Trigger recursion is handled appropriately
- [ ] All SOQL queries have been written as selectively as possible, and limited where appropriate
- [ ] Code is properly wrapped in try-catch and exceptions handled through the error handling framework
- [ ] This request includes batch or queueable apex
	- [ ] Batch apex is appropriately daisy-chained where required
- [ ] This request includes scheduable jobs
	- [ ] Scheduable jobs are properly injected into the scheduable framework

#### Tests
- [ ] Test coverage is > 85%
- [ ] Tests are in separate test classes
- [ ] Test conform to naming conventions
- [ ] All tests have positive and negative assertions
- [ ] Test data properly stands-up TDTM records where needed
- [ ] Tests make use of StartTest() and StopTest()
- [ ] Tests leverage different users and profiles

#### Visualforce
- [ ] This request includes VisualForce
	- [ ] Request meets mobile requirements
	- [ ] All pages are 508 compliant
	- [ ] All labels and text are translatable via Custom Labels
	- [ ] This VF has a controller
		- [ ] Controller sharing has been determined
		- [ ] Parameters are properly escaped
- [ ] CSRF protections are in place
- [ ] FLS is appropriately checked and enforced
- [ ] Business logic is not included in the controller
- [ ] Controller uses bind variables in SOQL
- [ ] Viewstate is managed through use of transients, etc.
	- [ ] Page includes custom CSS
	- [ ] Page include custom JS
		- [ ] Page includes new libraries or library requirements
		- [ ] Libraries are zipped and hosted in Static Resources
- [ ] Visualforce page utilizes existing VF components
- [ ] Templates are used where appropriate
- [ ] Visualforce page is an embedded component or page
- [ ] Visualforce page accepts user parameters for extendability
- [ ] User parameters are properly escaped and cleaned

**Comment by njjc** [comment]
**lurch: review apex visualforce

**Comment by LurchTheButler** [comment]
#### General
- [ ] All new pages, classes or metadata conform to naming conventions

#### Apex
- [ ] This request includes Apex
	- [ ] Sharing/Without/None has been explicitly determined and explained
	- [ ] FLS is appropriately checked and enforced
- [ ] Custom settings access is done through appropriate interfaces
- [ ] Methods are appropriately sized and scopes
- [ ] Inner classes utilized where appropriate
- [ ] Dependencies are done explicitly or have been noted where dynamic
- [ ] API Version is explicitly set to the latest or otherwise indicated as to why not
- [ ] Code is bulkified
	- [ ] No SOQL or DML in For loops
	- [ ] Async calls in For loops are properly rate limited
	- [ ] Code utilized DML wrapper where appropriate
- [ ] No new triggers have been added
- [ ] All currency operations are multi-currency capable and compliant
- [ ] Any new triggered functionality properly implements TDTM interface
	- [ ] A new TDTM record has been added to the install script for this class
	- [ ] Trigger recursion is handled appropriately
- [ ] All SOQL queries have been written as selectively as possible, and limited where appropriate
- [ ] Code is properly wrapped in try-catch and exceptions handled through the error handling framework
- [ ] This request includes batch or queueable apex
	- [ ] Batch apex is appropriately daisy-chained where required
- [ ] This request includes scheduable jobs
	- [ ] Scheduable jobs are properly injected into the scheduable framework

#### Tests
- [ ] Test coverage is > 85%
- [ ] Tests are in separate test classes
- [ ] Test conform to naming conventions
- [ ] All tests have positive and negative assertions
- [ ] Test data properly stands-up TDTM records where needed
- [ ] Tests make use of StartTest() and StopTest()
- [ ] Tests leverage different users and profiles

#### Visualforce
- [ ] This request includes VisualForce
	- [ ] Request meets mobile requirements
	- [ ] All pages are 508 compliant
	- [ ] All labels and text are translatable via Custom Labels
	- [ ] This VF has a controller
		- [ ] Controller sharing has been determined
		- [ ] Parameters are properly escaped
- [ ] CSRF protections are in place
- [ ] FLS is appropriately checked and enforced
- [ ] Business logic is not included in the controller
- [ ] Controller uses bind variables in SOQL
- [ ] Viewstate is managed through use of transients, etc.
	- [ ] Page includes custom CSS
	- [ ] Page include custom JS
		- [ ] Page includes new libraries or library requirements
		- [ ] Libraries are zipped and hosted in Static Resources
- [ ] Visualforce page utilizes existing VF components
- [ ] Templates are used where appropriate
- [ ] Visualforce page is an embedded component or page
- [ ] Visualforce page accepts user parameters for extendability
- [ ] User parameters are properly escaped and cleaned

**Comment by njjc** [comment]
@davidhabib this is ready to go. I forget where we left off after our last review, let me know if we should schedule a time to review this together!

File: src/classes/OPP_OpportunityContactRoles_TEST.cls [hunk 1]
```diff
@@ -34,31 +34,31 @@
 */
 @isTest
 private class OPP_OpportunityContactRoles_TEST {
+
+    /*********************************************************************************************************
+    * @description if you only want to run one test in this class, fill in its name here.
+    * if you want to run all tests, then use '*'
+    */
+    private static string strTestOnly = 'avoidDuplicateCRs';
```
Comment (davidhabib) [inline src/classes/OPP_OpportunityContactRoles_TEST.cls]: think you want this back to *!

Comment (njjc) [inline src/classes/OPP_OpportunityContactRoles_TEST.cls]: Blarg, I need some sort of automation to remind me of this :)

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 2]
```diff
@@ -322,10 +139,368 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
...
+                    if (newOpp.get(field)!=null)
+                        mapConIdContact.put((id)newOpp.get(field), null);
+                    if (oldOpp.get(field)!=null)
+                        mapConIdContact.put((id)oldOpp.get(field), null);
+                }
+            }
+        }
+
+        if (!mapConIdContact.isEmpty()) {
+            mapConIdContact = new map<id,Contact>([SELECT Id, Name, MailingAddress FROM Contact WHERE Id IN :mapConIdContact.keySet()]);
```
Comment (davidhabib) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: why include MailingAddress in this query?

**Comment by davidhabib** [comment]
OK, I reviewed the pull request, but I really got lost in all the changes in HH_OppContactRoles_TDTM and OPP_OpportunityContactRoles_TDTM, so I think I need you to walk me through it all.  thanks.

**Comment by njjc** [comment]
Don't merge this yet.

**Comment by njjc** [comment]
@davidhabib this is ready now.

**Pull request merged**